lib/Sema/CSSimplify.cpp

```diff
@@ -4888,9 +4888,31 @@ bool ConstraintSystem::repairFailures(
     break;
   }
 
+  // Accept mutable pointers in the place of immutables.
+  if (toImmutablePossible(lhs, rhs))
+    conversionsOrFixes.push_back(
+        ConversionRestrictionKind::PointerToPointer);
+
   return !conversionsOrFixes.empty();
 }
 
+bool ConstraintSystem::toImmutablePossible(Type lhs, Type rhs) {
+  if (rhs->isUnsafeRawPointer() && (lhs->isUnsafeMutableRawPointer() ||
+       lhs->isUnsafeMutablePointer() || lhs->isUnsafePointer()))
+    return true; // Unsafe[Mutable][Raw]Pointer -> UnsafeRawPointer
+
+  auto firstGenericArgument = [&](Type ty) -> CanType {
+    return dyn_cast<BoundGenericStructType>(ty->getCanonicalType())
+      ->getGenericArgs()[0]->getCanonicalType();
+  };
+
+  if (lhs->isUnsafeMutablePointer() && rhs->isUnsafePointer() &&
+      firstGenericArgument(lhs) == firstGenericArgument(rhs))
```

Comment:
Comparing canonical types isn't enough here, because `lhs` or `rhs` could contain type variables that haven't been determined yet. When the solver determines that it's possible to convert from mutable to immutable, you'll need to introduce a same-type constraint between the arguments. I suspect you'll be able to see the problem with a test like this:

```swift
func f<T>(_: UnsafePointer<T>, _: UnsafePointer<T>) { }

func g(mut: UnsafeMutablePointer<Int>, imm: UnsafePointer<Int>) {
  f(imm, mut)
  f(mut, imm)
}
```